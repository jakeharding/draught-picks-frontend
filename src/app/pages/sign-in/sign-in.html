<!--
  Generated template for the SignInPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <draught-header></draught-header>
</ion-header>

<ion-content class="ion-padding ion-text-center">
  <ion-row>
    <ion-col size="12" sizeSm="10" offsetSm="1" sizeMd="8" offsetMd="2" sizeLg="6" offsetLg="3" sizeXl="4" offsetXl="4">
      <h1>Please Sign In</h1>
      <form [formGroup]="signInForm" (submit)="signIn()">
        <ion-item>
          <ion-label position="floating">Username or Email</ion-label>
          <ion-input tid="usernameField" formControlName="username" type="text"></ion-input>
        </ion-item>
        <br>
        <ion-item>
          <ion-label position="floating">Password</ion-label>
          <ion-input tid="passwordField" formControlName="password" type="password"></ion-input>
        </ion-item>
<!--        Work around for ion-button not submitting form when enter key is pressed-->
        <button type="submit" class="ion-page-hidden"></button>
        <ion-button
            tid="signInButton"
            (click)="signIn();"
            class="ion-margin-top"
            [disabled]="signInForm.invalid"
        >Sign In</ion-button>
        <div class="ion-text-center">
          <p>Do you not have an account?</p>
          <a routerLink="/register" routerDirection="forward">Click here to register</a>
        </div>
      </form>
    </ion-col>
  </ion-row>
</ion-content>
