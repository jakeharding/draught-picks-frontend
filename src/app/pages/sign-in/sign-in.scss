:host {
  ion-title{
    position: absolute;
    top: 0;
    left: 0;
    padding: 0 90px 1px;
    width: 100%;
    height: 100%;
    text-align: center;
    font-weight: bold;
  }
  ion-content{
    text-align: center;
  }

  a{
    cursor: pointer;
    text-decoration: none;
  }
  form{
    text-align: center;
  }
}
