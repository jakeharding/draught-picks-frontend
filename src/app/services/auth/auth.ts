import { HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
import { Injectable, Injector } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import AuthResponse from '../../models/AuthResponse';

/**
 * Generated class for the AuthProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
 *
 * Used to authenticate the user
 */

export const TOKEN_STO_KEY = 'draughtPicksToken';


@Injectable({
  providedIn: 'root',
})
export class AuthProvider implements HttpInterceptor {

  signInUrl: string;

  constructor(public http: HttpClient) {
    this.signInUrl = `${Env.REST_API_ROOT}/login`;
  }

  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    if (this.isLoggedIn()) {
      request = request.clone({
        setHeaders: {
          Authorization: `Token ${this.getToken()}`
        }
      });
    }
    return next.handle(request);
  }

  signIn(creds: any): Promise<AuthResponse> {
    return this.http.post(this.signInUrl, creds).toPromise().then((resp) => {
      return resp as AuthResponse;
    });
  }

  getToken(): string {
    return localStorage.getItem(TOKEN_STO_KEY);
  }

  setToken(token: string) {
    localStorage.setItem(TOKEN_STO_KEY, token);
  }

  clearToken() {
    localStorage.removeItem(TOKEN_STO_KEY);
  }

  isLoggedIn() {
    return !!this.getToken();
  }
}

export function LoginRequired(target: any) {
  const authProvider = Injector.create([{provide: AuthProvider, deps: [] }]).get(AuthProvider);
  target.prototype.ionViewCanEnter = () => {
    if (!authProvider.isLoggedIn()) {
      location.assign('sign-in');
    }
    return authProvider.isLoggedIn();
  };
}
