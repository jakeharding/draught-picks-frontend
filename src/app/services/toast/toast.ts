import { Injectable } from '@angular/core';
import { ToastController } from '@ionic/angular';

/*
  Generated class for the ToastProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable({
  providedIn: 'root',
})
export class ToastProvider {
  static readonly defaultErrorMsg = 'We are having trouble connecting at the moment. ' +
  'Please check your network connection and try again.';
  private readonly errorToastClass = 'error-toast';
  private readonly successToastClass = 'success-toast';

  constructor(private toastCtrl: ToastController) {}

  errorToast(message: string = ToastProvider.defaultErrorMsg): void {
    this.createToast(message, this.errorToastClass);
  }

  successToast(message: string): void {
    this.createToast(message, this.successToastClass);
  }

  async createToast(message: string,
                    cssClass: string,
                    duration: number = 3000,
                    position: 'top' | 'bottom' | 'middle' = 'top'
              ) {
    const toast = await this.toastCtrl.create({ message, cssClass, duration, position });
    await toast.present();
  }

}
