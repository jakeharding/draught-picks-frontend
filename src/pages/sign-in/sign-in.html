<!--
  Generated template for the SignInPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<draught-header></draught-header>

<ion-content padding>
  <ion-row>
    <ion-col class="below-header" col-12 col-sm-10 offset-sm-1 col-md-8 offset-md-2 col-lg-6 offset-lg-3 col-xl-4 offset-xl-4>
      <h1>Please Sign In</h1>
      <form [formGroup]="signInForm">
        <ion-item>
          <ion-label floating>Username or Email</ion-label>
          <ion-input tid="usernameField" formControlName="username" type="text"></ion-input>
        </ion-item>
        <br>
        <ion-item>
          <ion-label floating>Password</ion-label>
          <ion-input tid="passwordField" formControlName="password" type="password"></ion-input>
        </ion-item>
        <button
            ion-button
            tid="signInButton"
            (click)="signIn();"
            margin-top
            [disabled]="signInForm.invalid"
        >Sign In</button>
        <p>Do you not have an account?</p>
        <a href="#/registration" tid="registrationLink">Click here to register</a>
      </form>
    </ion-col>
  </ion-row>
</ion-content>
