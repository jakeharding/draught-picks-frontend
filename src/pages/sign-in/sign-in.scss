page-sign-in {
  ion-title{
    position: absolute;
    top: 0;
    left: 0;
    padding: 0 90px 1px;
    width: 100%;
    height: 100%;
    text-align: center;
    font-weight: bold;
  }
  ion-content{
    text-align: center;
  }

//ion-input{
//    border:
//    width: 25%;
//    display: block;
//  }
  a{
    cursor: pointer;
    text-decoration: underline;
  }
  form{
    text-align: center;
  }
}
