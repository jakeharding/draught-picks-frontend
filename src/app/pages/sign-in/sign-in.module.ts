import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { SignInPage } from './sign-in';
import { ComponentsModule } from '../../components/components.module';
import { RouterModule } from '@angular/router';


@NgModule({
  declarations: [
    SignInPage,
  ],
  imports: [
    IonicModule,
    RouterModule.forChild([{ path: 'signiin', component: SignInPage}]),
    ComponentsModule,
  ],
})
export class SignInPageModule {}



