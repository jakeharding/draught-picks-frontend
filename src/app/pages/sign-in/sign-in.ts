import { NavController } from '@ionic/angular';
import { Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { RegistrationPage } from '../registration/registration';
import { AuthProvider } from '../../services/auth/auth';
import { BasePage } from '../BasePage';
import { ToastProvider } from '../../services/toast/toast';

/**
 * Generated class for the SignInPage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@Component({
  selector: 'page-sign-in',
  templateUrl: 'sign-in.html',
  styleUrls: ['./sign-in.scss']
})
export class SignInPage extends BasePage {
  signInForm: FormGroup;
  goToRegistration: any;
  goToResetPassword;
  private username: string;
  private password: string;

  constructor(public navCtrl: NavController, public formBuilder: FormBuilder, public toastProvider: ToastProvider,
              public authProvider: AuthProvider) {
    super('sign-in');
    if (this.authProvider.isLoggedIn()) {
      this.navCtrl.navigateRoot('tabs');
    }
    // this.goToRegistration = RegistrationPage;
    this.signInForm = formBuilder.group({
      username: [this.username, Validators.compose([Validators.required])],
      password: [this.password, Validators.compose([Validators.required, Validators.minLength(4)])]
    });
  }

  /**
   * signIn function
   * No Parameters
   * Gets the username and password values and validates the information then sends
   * user to the home page
   */
  public signIn() {
    this.authProvider.signIn(this.signInForm.value).then(
      (response) => {
        this.authProvider.setToken(response.token);
        this.navCtrl.navigateRoot('/tabs/home');
      },
      () => {
        this.toastProvider.errorToast('Unable to sign you in. Have you verified your email address?');
      }
    );

  }
}

