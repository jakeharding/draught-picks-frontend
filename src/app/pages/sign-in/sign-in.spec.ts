import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { SignInPage } from './sign-in';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NavController, ToastController } from '@ionic/angular';
import { FormBuilder, FormGroupDirective } from '@angular/forms';
import { AuthProvider } from '../../services/auth/auth';
import { mockNavController, mockToastProvider } from '../../../../setup-jest';
import { ToastProvider } from '../../services/toast/toast';
import { RouterTestingModule } from '@angular/router/testing';

/**
 * sign-in.spec.ts
 *
 * Created by jake
 * Created on 2018-11-21
 *
 * Test SignInPage.
 */

describe('SignInPage', () => {
  let fixture: ComponentFixture<SignInPage>;
  let component: SignInPage;

  const mockAuthProvider = {
    isLoggedIn: jest.fn().mockReturnValue(true),
    signIn: jest.fn().mockReturnValue(Promise.resolve({token: 'token'})),
    setToken: jest.fn()
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [SignInPage, FormGroupDirective],
      imports: [RouterTestingModule],
      providers: [
        { provide: ToastProvider, useValue: mockToastProvider},
        { provide: AuthProvider, useValue: mockAuthProvider},
        { provide: NavController, useValue: mockNavController},
        FormBuilder
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SignInPage);
    component = fixture.debugElement.componentInstance;
  });

  it('should initialize the SignInPage and set root to TabsPage if logged in', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
    expect(mockNavController.navigateRoot).toHaveBeenCalledTimes(1);
    expect(mockNavController.navigateRoot).toHaveBeenCalledWith('tabs');
  });

  test('signIn should call authProvider.setToken and navCtrl.navigateRoot when promise resolves', async () => {
    jest.clearAllMocks();
    await component.signIn();
    expect(mockAuthProvider.setToken).toHaveBeenCalledTimes(1);
    expect(mockAuthProvider.setToken).toHaveBeenCalledWith('token');
    expect(mockNavController.navigateRoot).toHaveBeenCalledTimes(1);
    expect(mockNavController.navigateRoot).toHaveBeenCalledWith('/tabs/home');
  });

  test('signIn should show error toast when promise is rejected', async () => {
    mockAuthProvider.signIn.mockReturnValue(Promise.reject('error'));
    await component.signIn();
    expect(mockToastProvider.errorToast).toHaveBeenCalledTimes(1);
    expect(mockToastProvider.errorToast).toHaveBeenCalledWith('Unable to sign you in. Have you verified your email address?');
  });
});
