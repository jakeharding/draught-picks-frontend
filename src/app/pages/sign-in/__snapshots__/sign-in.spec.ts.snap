// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInPage should initialize the SignInPage and set root to TabsPage if logged in 1`] = `
<page-sign-in
  authProvider={[Function Object]}
  formBuilder={[Function FormBuilder]}
  navCtrl={[Function Object]}
  signInForm={[Function FormGroup]}
  toastProvider={[Function Object]}
>
  <draught-header /><ion-content
    padding=""
  >
    <ion-row>
      <ion-col
        class="below-header"
        col-12=""
        col-lg-6=""
        col-md-8=""
        col-sm-10=""
        col-xl-4=""
        offset-lg-3=""
        offset-md-2=""
        offset-sm-1=""
        offset-xl-4=""
      >
        <h1>
          Please Sign In
        </h1>
        <form>
          <ion-item>
            <ion-label
              floating=""
            >
              Username or Email
            </ion-label>
            <ion-input
              formcontrolname="username"
              tid="usernameField"
              type="text"
            />
          </ion-item>
          <br />
          <ion-item>
            <ion-label
              floating=""
            >
              Password
            </ion-label>
            <ion-input
              formcontrolname="password"
              tid="passwordField"
              type="password"
            />
          </ion-item>
          <button
            ion-button=""
            margin-top=""
            tid="signInButton"
          >
            Sign In
          </button>
          <p>
            Do you not have an account?
          </p>
          <a
            routerdirection="root"
            routerlink="register"
          >
            Click here to register
          </a>
        </form>
      </ion-col>
    </ion-row>
  </ion-content>
</page-sign-in>
`;
